test/integration/health_check_integration_test.cc

```diff
@@ -319,9 +319,20 @@ TEST_P(HttpHealthCheckIntegrationTest, SingleEndpointGoAwayError) {
   test_server_->waitForCounterGe("cluster.cluster_1.health_check.failure", 1);
   EXPECT_EQ(0, test_server_->counter("cluster.cluster_1.health_check.success")->value());
   EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
-  test_server_->waitForCounterGe("cluster.cluster_1.health_check.goaway_error", 1);
 
-  // Advance time to cause another health check.
+  // Now we want to advance time to cause another health check. Ideally the
+  // waitForDisconnect() step above would guarantee that the health checker has
+  // seen the GOAWAY, closed the health check connection, and finally enabled
+  // the health check interval timer so we can advance simulated time to cause
+  // another check. However, due to special handling for protocol errors, there
+  // is a race where both the downstream (health checker) and upstream (fake
+  // upstream) close the connection, so we may exit waitForDisconnect() before
+  // the health checker has actually seen the GOAWAY and enabled the timer. To
+  // address this, we use a small real sleep before advancing thee simulated
+  // time system to give the health checker time to catch up. If we advance time
+  // before the health check timer is enabled, the next health check will never
+  // happen.
+  timeSystem().realSleepDoNotUseWithoutScrutiny(std::chrono::milliseconds(100));
```

Comment:
Sorry to keep going back and forth on this, but maybe we should talk on Slack. It's not OK to use a real sleep here since it will flake. Please either stop using simulated time entirely, or just do what we talked about before and wait on the stat. It's not clear to me that simulated time is the best use for this case? Since you are just waiting for an HC failure to occur and then another reconnect?

/wait